src/inspector_io.cc

```diff
@@ -0,0 +1,439 @@
+}
+
+bool InspectorIo::IsStarted() {
+  return !!platform_;
+}
+
+void InspectorIo::WaitForDisconnect() {
+  if (state_ == State::kConnected) {
+    shutting_down_ = true;
+    Write(TransportAction::kStop, 0, StringView());
+    fprintf(stderr, "Waiting for the debugger to disconnect...\n");
+    fflush(stderr);
+    parent_env_->inspector_agent()->RunMessageLoop();
+  }
+}
+
+#define READONLY_PROPERTY(obj, str, var)                                      \
+  do {                                                                        \
+    obj->DefineOwnProperty(env->context(),                                    \
+                           OneByteString(env->isolate(), str),                \
+                           var,                                               \
+                           v8::ReadOnly).FromJust();                          \
+  } while (0)
+
+// static
+void InspectorIo::ThreadCbIO(void* io) {
+  static_cast<InspectorIo*>(io)->WorkerRunIO<InspectorSocketServer>();
+}
+
+// static
+template <typename Transport>
+void InspectorIo::WriteCbIO(uv_async_t* async) {
+  TransportAndIo<Transport>* io_and_transport =
+      static_cast<TransportAndIo<Transport>*>(async->data);
+  if (io_and_transport == nullptr) {
+    return;
+  }
+  MessageQueue<TransportAction> outgoing_messages;
+  InspectorIo* io = io_and_transport->second;
+  io->SwapBehindLock(&io->outgoing_message_queue_, &outgoing_messages);
+  for (const auto& outgoing : outgoing_messages) {
+    switch (std::get<0>(outgoing)) {
+    case TransportAction::kStop:
+      io_and_transport->first->Stop(nullptr);
+      break;
+    case TransportAction::kSendMessage:
+      std::string message = StringViewToUtf8(std::get<2>(outgoing)->string());
+      io_and_transport->first->Send(std::get<1>(outgoing), message);
+      break;
+    }
+  }
+}
+
+template<typename Transport>
+void InspectorIo::WorkerRunIO() {
+  uv_loop_t loop;
+  int err = uv_loop_init(&loop);
+  CHECK_EQ(err, 0);
+  io_thread_req_.data = nullptr;
+  err = uv_async_init(&loop, &io_thread_req_, WriteCbIO<Transport>);
+  CHECK_EQ(err, 0);
+  std::string script_path;
+  if (!script_name_.empty()) {
+    uv_fs_t req;
+    if (0 == uv_fs_realpath(&loop, &req, script_name_.c_str(), nullptr))
+      script_path = std::string(reinterpret_cast<char*>(req.ptr));
+    uv_fs_req_cleanup(&req);
+  }
+  InspectorIoDelegate delegate(this, script_path, script_name_,
+                               options_.wait_for_connect());
+  delegate_ = &delegate;
+  InspectorSocketServer server(&delegate,
+                               options_.host_name(),
+                               options_.port());
+  TransportAndIo<Transport> queue_transport(&server, this);
+  io_thread_req_.data = &queue_transport;
+  if (!server.Start(&loop)) {
+    state_ = State::kError;  // Safe, main thread is waiting on semaphore
+    uv_close(reinterpret_cast<uv_handle_t*>(&io_thread_req_), nullptr);
+    uv_loop_close(&loop);
```

Comment:
Can you `CHECK_EQ(0, uv_loop_close(&loop))`?  I'm pretty sure this won't work on account of the closing handle.